core/src/backend/ui.rs

```diff
@@ -1,6 +1,47 @@
+use crate::backend::navigator::OwnedFuture;
 use crate::events::{KeyCode, PlayerEvent};
+pub use crate::loader::Error as LoaderError;
+use chrono::{DateTime, Utc};
+use downcast_rs::Downcast;
 use std::collections::HashSet;
 
+/// A filter specifying a category that can be selected from a file chooser dialog
+pub struct FileFilter {
+    /// The description of the catagory
+    pub description: String,
+    /// A semicolon ';' delimited list of acceptable windows file extensions that can be selected
+    /// in this category, with a */wildcard before each extension
+    pub extensions: String,
+    /// A semicolon ';' delimited list of acceptable MacOs file extensions that can be selected in
+    /// this category, with a */wildcard before each extension
+    /// Note that a list of file filters will either all have Some(_) mac_type or all will have None
+    pub mac_type: Option<String>,
+}
+
+/// A result of a file selection
+pub trait FileDialogResult: Downcast {
```

Comment:
Instead of defining `FileDialogResult` as a trait, I think it can be made a `struct` with such fields. Then each UI backend will need to convert it's native representation to a `FileDialogResult` before returning it. It just seems to me much more simple this way.

Reply:
I do agree with this in part, but I'm not sure how to handle `refresh` and `write`. Writing could just have a platform check, and be a noop on web but I feel handling platform specifics are better left to each platforms respective backend. Refreshing has the same issue, in that it would have to be aware of how to re-read the source file and metadata on each platform.